clients/src/main/java/org/apache/kafka/common/utils/AclUtils.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.common.utils;
+
+import org.apache.kafka.common.resource.Resource;
+import org.apache.kafka.common.resource.ResourceFilter;
+import org.apache.kafka.common.resource.ResourceNameType;
+import org.apache.kafka.common.resource.ResourceType;
+
+public final class AclUtils {
+    public static final ResourceNameType LITERAL = ResourceNameType.LITERAL;
+    public static final String WILDCARD_RESOURCE = "*";
+
+    private AclUtils() {
+    }
+
+    /**
+     * Checks for a match between a resource path under which ACLs are stored and a resource filter.
+     *
+     * @param aclPath the resource path under which ACLs are stored.
+     * @param filter  a filter to match against.
+     * @return {@code true} if the resource matches the filter, {@code false} otherwise.
+     */
+    public static boolean matchResource(final Resource aclPath, final ResourceFilter filter) {
+        throwOnInvalidParams(aclPath);
+
+        if (!filter.resourceType().equals(ResourceType.ANY) && !filter.resourceType().equals(aclPath.resourceType())) {
+            return false;
+        }
+
+        if (!filter.nameType().equals(ResourceNameType.ANY) && !filter.nameType().equals(aclPath.resourceNameType())) {
+            return false;
+        }
+
+        if (filter.name() == null) {
+            return true;
+        }
+
+        if (filter.nameType().equals(aclPath.resourceNameType())) {
+            return aclPath.name().equals(filter.name());
+        }
+
+        switch (aclPath.resourceNameType()) {
+            case LITERAL:
+                return aclPath.name().equals(filter.name()) || aclPath.name().equals(WILDCARD_RESOURCE);
+
+            case PREFIXED:
```

Comment:
This is wrong.  AclFilter objects should match exactly or not at all.  Trying to remove a wildcard AclBinding should not remove every AclBinding.  Trying to remove a prefix AclBinding should not remove every AclBinding with that prefix.

Reply:
This KIP / PR has to see the semantics of matching a filter to a resource change.  Please see https://github.com/apache/kafka/pull/5117#discussion_r192668669

Reply:
Sorry, but this is wrong.  There are two "semantics of matching".
1. The semantics of an AclBindingFilter matching an AclBinding
2. The semantics of an AclBinding matching an ACL

Prefix comparisons are only relevant to #2, not to #1.

When I try to remove an AclBinding for a prefix ACL beginning with "foo", I don't want to remove ACLs the start with foobar, foo2, etc.  When I remove a wildcard ACL, I don't want to remove all ACLs.  And so forth.

Reply:
To summarize our offline discussion, the proposal is to override ResourceNameType.ANY in a somewhat odd, but potentially useful way.  Let's pick up this discussion in a follow on PR.